system/include/emscripten/atomic.h

```diff
@@ -19,63 +19,63 @@ extern "C" {
 
 // Atomically stores the given value to the memory location, and returns the
 // value that was there prior to the store.
-uint8_t emscripten_atomic_exchange_u8(void/*uint8_t*/ *addr, uint8_t newVal);
-uint16_t emscripten_atomic_exchange_u16(void/*uint16_t*/ *addr, uint16_t newVal);
-uint32_t emscripten_atomic_exchange_u32(void/*uint32_t*/ *addr, uint32_t newVal);
-uint64_t emscripten_atomic_exchange_u64(void/*uint64_t*/ *addr, uint64_t newVal);
+uint8_t emscripten_atomic_exchange_u8(void/*uint8_t*/ *addr __attribute__((nonnull)), uint8_t newVal);
```

Comment:
Wouldn't it also be nice to have the compiler warning if a user passes NULL here?

Reply:
The compiler will warn if a user passes a null here -  that is what `__attribute__((nonnull))` will do. But in addition to issuing the warning, the compiler can optimize code inside the function, and at the calls to that function with the expectation that any non-statically-inferrable value passed here will not be a null.

So i.e. if one had silly code like

```c
void foo(const char *x __attribute__((nonnull)))
{
	printf("%s\n", x);
}

extern const char *bar(void);

int main()
{
	const char *s = bar();
	foo(s);
	if (!s) // This can be assumed never to be true, since the contract with foo says it can only be called with nonnull pointers. Clang could issue a warning here.
	{
		// we cannot get here!
		printf("s was null!\n");
	}
}
```
However now I see that Clang does not actually do this kind of analysis when trying out in practice. :/ 

To answer the difference between `__attribute__((nonnull))` and `_Nonnull`: if instead the function `foo` read

```c
void foo(const char *x __attribute__((nonnull)))
{
    if (x)
	printf("%s\n", x);
}
```

the compiler will issue a diagnostic warning, because `if (x)` is tautologically true: the contract with `x` says it will never be null.

So if we have a function that does want to do something different when `x` actually is null, then we want to just use the weaker `_Nonnull` semantics: 

```c
void foo(const char * _Nonnull x) // issue a warning if x is null, but don't affect codegen
{
    if (x)
	printf("%s\n", x);
}
```

Reply:
I see, from your earlier description had though that one triggered optimization and the triggers warnings, but its seem that the both trigger warning and in addition the the `__attribute__` form allows optimization.

This makes sense to me now.